websites/G/Granblue Fantasy/presence.ts

```diff
@@ -5,65 +5,151 @@ const presence = new Presence({
+		[Elements.Earth]:
+			"https://gbf.wiki/images/thumb/9/94/Icon_Element_Earth.png/36px-Icon_Element_Earth.png",
+		[Elements.Wind]:
+			"https://gbf.wiki/images/thumb/6/62/Icon_Element_Wind.png/20px-Icon_Element_Wind.png",
+		[Elements.Light]:
+			"https://gbf.wiki/images/thumb/1/19/Icon_Element_Light.png/20px-Icon_Element_Light.png",
+		[Elements.Dark]:
+			"https://gbf.wiki/images/thumb/a/a5/Icon_Element_Dark.png/20px-Icon_Element_Dark.png",
+	},
+	ElementsNames = {
+		[Elements.Plain]: "Plain",
+		[Elements.Fire]: "Fire",
+		[Elements.Water]: "Water",
+		[Elements.Earth]: "Earth",
+		[Elements.Wind]: "Wind",
+		[Elements.Light]: "Light",
+		[Elements.Dark]: "Dark",
+	};
+
 interface GameStatus {
+	turn: number;
+	battle: {
+		count: number;
+		total: number;
+	};
 	boss: {
 		param: {
 			hp: string;
 			hpmax: string;
 			alive: 1 | 0;
+			name: {
+				ja: string;
+				en: string;
+			};
+		}[];
+	};
+	player: {
+		param: {
+			attr: Elements;
+			alive: 1 | 0;
+			leader: 1 | 0;
+			name: string;
+			hp: number;
+			hpmax: number;
+			pid: string;
 		}[];
 	};
 }
+interface UserData {
+	baseUri: string;
+	imgUri: string;
+	rank: number;
+	id: number;
+}
 
-let gameStatus: GameStatus;
+let gameStatus: GameStatus, userData: UserData;
 
-// Decreasing latency may greatly impact on gameplay/performance
 script.id = eventId;
 script.appendChild(
 	document.createTextNode(`
-  let isRunning = false;
-  setInterval(() => {
-    if (isRunning) return;
-    isRunning = true;
-    const getCircularReplacer = () => {
-      const seen = new WeakSet();
-      return (key, value) => {
-        if (typeof value === "object" && value !== null) {
-          if (seen.has(value)) return;
-          else seen.add(value);
-        }
-        return value;
-      };
-    };
-    const gGameStatus = JSON.stringify(window.stage?.gGameStatus, getCircularReplacer());
-    const pmdEvent = new CustomEvent("${eventId}", {
-      detail: {
-        gGameStatus
-      }
-    });
-    window.dispatchEvent(pmdEvent);
-    isRunning = false;
-  }, 4500);
+  	let isRunning = false;
+  	setInterval(() => {
+		if (isRunning) return;
+		isRunning = true;
+		try {
+			const gGameStatus = JSON.stringify({
+				turn: window.stage?.gGameStatus.turn,
+				battle: window.stage?.pJsnData.battle,
+				boss: window.stage?.gGameStatus.boss,
+				player: window.stage?.pJsnData.player
+			});
+			const userData = JSON.stringify({
+				rank: window.Game?.userRank,
+				id: window.Game?.userId,
+				baseUri: window.Game?.baseUri,
+				imgUri: window.Game?.imgUri
+			});
+			const pmdEvent = new CustomEvent("${eventId}", {
+				detail: {
+					gGameStatus,
+					userData
+				}
+			});
+			window.dispatchEvent(pmdEvent);
+			isRunning = false;
+		} catch (error) {
+			window.dispatchEvent(new CustomEvent("${eventId}", {
+				detail: {
+					gGameStatus: null,
+					userData: null,
+					error: error.message,
+					stack: error.stack
+				}
+			}));
+			isRunning = false;
+		}
+  	}, 3000);
```

Comment:
Try using `getPageVariable`

Reply:
All these variables have circular dependencies, which makes the browser absolutely shits itself when `JSON.stringify` is called on it, which is also what the extension is using to send variable data.

This is why the script was used in the first place

Reply:
References to the reasons behind the script:
https://discord.com/channels/493130730549805057/607524579874832446/930332649795493938
https://discord.com/channels/493130730549805057/607524579874832446/930390425481662504

Reply:
@Timeraa can confirm, but I am pretty sure the circular will be replaced with `"[Circular]"`

Reply:
> @Timeraa can confirm, but I am pretty sure the circular will be replaced with `"[Circular]"`

Unless you have manually implemented the circular replacer from here, which matches the string replacement you mentioned, then yes
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value

From what I'm seeing in the current stable version of the extension (`v2.7.4`), it does not, and `JSON.stringify` does not handle these cases by default.

Regardless, I have tested it and it does *still* freezes the browser tab

Reply:
There is also the issue of performance (as the stringified object is 20kb in size), which is why the script had to be run on a 3-4s delay `setInterval`.
This PR aims to reduce that by selecting only the few properties needed in the object, of which I don't think `getPageVariable` is capable of atm.

Reply:
I mean you can use dot props in the getPageVariables, so you only need to fetch what you need, and you can fetch multiple at once.

Reply:
For example in youtube: https://github.com/PreMiD/Presences/blob/9f824827e086587b3db9b512ff14d7c29a580436/websites/Y/YouTube/video_sources/api.ts#L28

Reply:
Gave it a shot, it would still run into some kind of issues that blocks the extension from updating, most likely XSS/circular dependency as I've encountered using the custom script

![image](https://github.com/user-attachments/assets/ed12fe5d-718b-48e7-b546-9932fa07452d)

Reply:
I somehow made that work, doesn't seem to have that much overhead either

Reply:
We do replace circular references with [Circular], this only applies to the `getPageVariable` method though.